design.md

```diff
@@ -0,0 +1,107 @@
+## Design Document - v0
+
+This is an oversimplified version of a kademlia/mainline DHT.
+
+For at least the v0 of our DHT we only care about storing and retrieving information about other DHT clients/servers; which we will call Peers.
+
+Mainline separates peers from Peers as one is the DHT client/server and the other is a BitTorrent provider.  
+We do not care about anything more than the DHT clients/servers for now, so this should simplify things a bit.
+
+## Terminology
+
+* `peer` - A participating client/server in the DHT network.
+* `k-buckets` - The routing table keeps a list of known GOOD peers.
+* `alpha` - A small number representing the degree of parallelism in network calls, usually 3.
+* `B` - The size in bits of the keys used to identify peers; in Kademlia this is 160, the length of a SHA1 digest.
+* `k` - The maximum number of contacts stored in a bucket; this is 20 in kademlia and 8 in mainline.
+* `tRefresh` - The time after which an otherwise unaccessed bucket must be refreshed; 3600s in Kademlia.
+* `refresh` - The process of getting fresh peers into unaccessed buckets.
+
+## Peer
+
+Each peer has an ID, for the purpose of this document we assume that the each ID is unique in the network.
+
+Kademlia uses 160-bit identifiers (SHA1) for peer IDs, we could explore the possibility of moving to some form of multihash at some point if we decide that this DHT does not have to be compatible with Kad.
```

Comment:
The assumption for now is that peer id will be the fingerprint of a public key.

This way when the connection to/from a peer is being established the peers (via TLS) will be able to exchange keys both for securing the connection and also to validate the validity of the peer id.

This should solve the issue of the spartan attack for the purposes of the DHT being used for peer discovery.